File: 4346-subsequence-after-one-replacement/subsequence-after-one-replacement.cpp
class Solution {
public:
    bool canMakeSubsequence(string s, string t) {
        int n = s.size();

        int i = 0;
        int j = 0;

        for (char ch : t) {
            if (j < n && s[j] == ch)
                j++;

            if (i < n && s[i] == ch)
                i++;
            else
                j = max(j, i + 1);

            if (i == n || j == n)
                return true;
        }

        return false;
    }
};

File: 4346-subsequence-after-one-replacement/subsequence-after-one-replacement.rs
impl Solution {
    pub fn can_make_subsequence(s: String, t: String) -> bool {
        let s = s.as_bytes();
        let n = s.len();

        let mut i = 0; // Next index in s to match (no replacement)
        let mut j = 0; // Next index in s to match (at most one replacement)

        for ch in t.bytes() {
            if j < n && s[j] == ch {
                j += 1;
            }

            if i < n && s[i] == ch {
                i += 1;
            } else {
                j = j.max(i + 1);
            }

            if i == n || j == n {
                return true;
            }
        }

        false
    }
}